### PR TITLE
Use middleware to add nonce and CSP headers

### PR DESCRIPTION
**Pull request by hursey013**
## Changes proposed in this pull request:

- Refactors middleware to allow chaining of multiple middlewares
- Refactors `Next/Image` to remove inline styles
- Add CSP headers and nonce via middleware

### Related issues

https://www.58bits.com/blog/chaining-or-combining-nextjs-middleware
https://nextjs.org/docs/app/building-your-application/configuring/content-security-policy
https://github.com/vercel/next.js/tree/canary/examples/with-strict-csp
https://github.com/vercel/next.js/issues/45184

### Submitter checklist

- [ ] Added logging is not capturing sensitive data and is set to an appropriate level (DEBUG vs INFO etc)
- [ ] Updated relevant documentation (README, ADRs, explainers, diagrams)

## Security considerations

None

### CODE REVIEW

File: src/components/Image.tsx [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import NextImage, { getImageProps } from 'next/image';
+import { ComponentProps } from 'react';
+
+export default function Image(props: ComponentProps<typeof NextImage>) {
+  const { props: nextProps } = getImageProps({
+    ...props,
+  });
+
+  // eslint-disable-next-line no-unused-vars
+  const { style: _omit, ...delegated } = nextProps;
```
Comment (echappen) [inline src/components/Image.tsx:10]: [question] Is this equivalent to `const { style, ...delegated } = nextProps`, i.e. the `style` key/value is being removed from the object?

File: src/components/Overlays/OverlayHeaderUsername.tsx [hunk 2]
```diff
@@ -13,10 +13,7 @@ export function OverlayHeaderUsername({
 }) {
   return (
     <>
-      <h2
-        className="margin-top-0 margin-bottom-7 text-uppercase text-light underline-base-light text-underline font-sans-xs"
-        style={{ textUnderlineOffset: '0.7em' }}
-      >
+      <h2 className="margin-top-0 margin-bottom-7 text-uppercase text-light underline-base-light text-underline font-sans-xs">
```
Comment (echappen) [inline src/components/Overlays/OverlayHeaderUsername.tsx:16]: [blocking question] Does the text underline offset style get added back somewhere?

File: src/middlewares/withAuth.ts [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+export function redirectToLogin(request: NextRequest): NextResponse {
+  const loginPath = logInPath();
+  const response = NextResponse.redirect(new URL(loginPath, request.url));
+  response.cookies.set('last_page', request.nextUrl.pathname);
+  return response;
+}
+
+export async function authenticateRoute(request: NextRequest) {
+  let response = NextResponse.next();
```
Comment (echappen) [inline src/middlewares/withAuth.ts:133]: [blocking question] My understanding of the stacked approach is that the same `response` object gets passed through and is modified by each middleware function. Does how we produce a response in each function matter? I see that this function uses `response = NextResponse.next()` whereas withCSP uses `response = await next(request, _next)`. Is one preferable over the other?

File: src/components/Image.tsx [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+  return <img {...delegated} />;
+}
```
Comment (echappen) [inline src/components/Image.tsx:13]: [blocking question] I see we're handling inline styles for images. Are the inline styles for the `ProgressBar` component affected at all by the new CSP headers?

**Review by echappen — COMMENTED** [review]
Thanks for all of this work, Brian, I really like the separation of concerns this approach takes. There are some concepts which I'd like to better understand before I approve. LMK if a chat would be better.

File: src/components/Image.tsx (hunk 1 above)
Comment (hursey013) [reply src/components/Image.tsx:10]: Yeah, I was following the workaround provided on a [related issue](https://github.com/vercel/next.js/issues/45184#issuecomment-1988319088), but now that I'm looking at it I'm not sure why `_omit` would be necessary, I can update to remove that.

File: src/components/Overlays/OverlayHeaderUsername.tsx (hunk 2 above)
Comment (hursey013) [reply src/components/Overlays/OverlayHeaderUsername.tsx:16]: Sorry, I pulled this out as I was experimenting. Can this be pulled into a reusable class as opposed to an inline style?

File: src/components/Image.tsx (hunk 1 above)
Comment (echappen) [reply src/components/Image.tsx:10]: No need to change, was just curious. I kind of like the explicitness of `_omit`.

File: src/components/Overlays/OverlayHeaderUsername.tsx (hunk 2 above)
Comment (echappen) [reply src/components/Overlays/OverlayHeaderUsername.tsx:16]: I think so. I doubt we'll have multiple types of text underline offset, so maybe something like:
```
.text-underline-offset { text-underline-offset: 0.7em; }
```

File: src/components/Image.tsx (hunk 4 above)
Comment (hursey013) [reply src/components/Image.tsx:13]: I believe we'll be able to grab the nonce off the headers and add it directly to the element using inline styles.

Comment (echappen) [reply src/components/Image.tsx:13]: As this is a breaking change for the progress bar, could you show how you would add the nonce to the progress bar component in this PR?

File: src/middlewares/withAuth.ts (hunk 3 above)
Comment (echappen) [reply src/middlewares/withAuth.ts:133]: We'll test this when deployed

**Review by echappen — APPROVED** [review]

**Pull request merged**